src/transformers/trainer.py

```diff
@@ -847,6 +847,15 @@ def create_optimizer(self):
                 )
             else:
                 self.optimizer = optimizer_cls(optimizer_grouped_parameters, **optimizer_kwargs)
+                if optimizer_cls.__name__ == "Adam8bit":
+                    import bitsandbytes
+
+                    manager = bitsandbytes.optim.GlobalOptimManager.get_instance()
+
+                    for module in self.model.modules():
+                        if isinstance(module, nn.Embedding):
```

Comment:
This is tricky here, since ideally we only want to switch to fp32 the word embeddings, but positional embeddings are a subclass of `nn.Embedding` as well.

Perhaps since positional embeddings are much smaller it's OK if they remain in fp32?

@TimDettmers, what do you think?

Reply:
Positional embedding can stay in fp32. I do not think positional embeddings make a big difference, but its important for pretraining models to have the embeddings in fp32.